ui/perfherder/graphs/GraphsContainer.jsx

```diff
@@ -1,15 +1,484 @@
+    top: point.y - yOffset,
+  });
+
+  showTooltip = (dataPoint, lock) => {
+    const position = this.getTooltipPosition(dataPoint);
+    this.hideTooltip.cancel();
+    this.tooltip.current.style.cssText = `left: ${position.left}px; top: ${position.top}px;`;
+
+    this.setState({
+      showTooltip: true,
+      lockTooltip: lock,
+      dataPoint,
+    });
+  };
+
+  setTooltip = (dataPoint, lock = false) => {
+    const { lockTooltip } = this.state;
+    const { updateStateParams } = this.props;
+
+    // we don't want the mouseOver event to reposition the tooltip
+    if (lockTooltip && !lock) {
+      return;
+    }
+    this.showTooltip(dataPoint, lock);
+
+    if (lock) {
+      updateStateParams({
+        selectedDataPoint: {
+          signature_id: dataPoint.datum.signature_id,
+          pushId: dataPoint.datum.pushId,
+          x: dataPoint.x,
+          y: dataPoint.y,
+        },
+      });
+    }
+  };
+
+  closeTooltip = () => {
+    this.setState({
+      showTooltip: false,
+      lockTooltip: false,
+      dataPoint: null,
+    });
+    this.props.updateStateParams({ selectedDataPoint: null });
+  };
+
+  // The Victory library doesn't provide a way of dynamically setting the left
+  // padding for the y axis tick labels, so this is a workaround (setting state
+  // doesn't work with this callback, which is why a class property is used instead)
+  setLeftPadding = (tick, index, ticks) => {
+    const highestTickLength = ticks[ticks.length - 1].toString();
+    this.leftChartPadding = highestTickLength.length * 8 + 10;
+    const numberFormat = new Intl.NumberFormat();
+
+    return numberFormat.format(tick);
+  };
+
+  updateData(zoom) {
+    const { testData } = this.props;
+
+    // we do this along with debouncing updateZoom to make zooming
+    // faster by removing unneeded data points based on the updated x,y
+    if (zoom.x && zoom.y) {
+      const scatterPlotData = testData
+        .flatMap(item => (item.visible ? item.data : []))
+        .filter(
+          data =>
+            data.x >= zoom.x[0] &&
+            data.x <= zoom.x[1] &&
+            data.y >= zoom.y[0] &&
+            data.y <= zoom.y[1],
+        );
+      this.setState({ scatterPlotData });
+    }
+  }
+
+  // debounced
+  hideTooltip() {
+    const { showTooltip, lockTooltip } = this.state;
+
+    if (showTooltip && !lockTooltip) {
+      this.setState({ showTooltip: false });
+    }
+  }
+
+  // debounced
+  updateZoom(zoom) {
+    const { showTooltip, lockTooltip } = this.state;
+
+    if (showTooltip && lockTooltip) {
+      this.closeTooltip();
+    }
+
+    this.props.updateStateParams({ zoom });
+    this.updateData(zoom);
+  }
+
+  render() {
+    const { testData, zoom, highlightedRevisions } = this.props;
+    const {
+      highlights,
+      scatterPlotData,
+      entireDomain,
+      showTooltip,
+      lockTooltip,
+      dataPoint,
+    } = this.state;
+
+    const highlightPoints = Boolean(highlights.length);
```

Comment:
You can also use a shortcut for things like this: ``!!highlights.length`` which will turn it into a boolean.